### PR TITLE
Initial search parameters / restriction, more logging

### PR DESCRIPTION
**Pull request by AlexReimann**
* Restrict the initial global constraint search
* Only search locally for global constraint search
* Parameterized parameters for global search window size
* Log origin and target info for global constraint search
* Log all inter trajectory constraints
* Log submap ids of dropped / succeeded

### CODE REVIEW

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 1]
```diff
@@ -268,7 +268,8 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
 
     bool global_search_time = node_time >= last_connection_time +
             common::FromSeconds(options_.global_constraint_search_after_n_seconds());
-    bool global_search = last_connection_time == common::Time() ||
+    bool initialization = last_connection_time == common::Time() && node_id.node_index <= 2;
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: const

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 2]
```diff
@@ -268,7 +268,8 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
...
+    bool global_search = initialization ||
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: const

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 3]
```diff
@@ -289,6 +290,15 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
                                            constant_data, global_node_pose,
                                            global_submap_pose);
   } else if (maybe_add_global_constraint) {
+    float distance = (global_node_pose.translation()- global_submap_pose.translation()).norm();
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: const

File: cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc [hunk 4]
```diff
@@ -187,9 +188,11 @@ FastCorrelativeScanMatcher3D::MatchWithSearchParameters(
   const std::vector<Candidate3D> lowest_resolution_candidates =
       ComputeLowestResolutionCandidates(search_parameters, discrete_scans);
 
+  std::string search_type = full ? "full search" : "local search";
```
Comment (johuber) [inline cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc]: const

File: cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc [hunk 5]
```diff
@@ -187,9 +188,11 @@ FastCorrelativeScanMatcher3D::MatchWithSearchParameters(
...
+
   u_int low_res_candidate_warn_threshold = 500000;
```
Comment (johuber) [inline cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc:206]: const

File: cartographer/mapping/internal/constraints/constraint_builder_3d.cc [hunk 6]
```diff
@@ -228,8 +228,8 @@ void ConstraintBuilder3D::ComputeConstraint(
     }
   } else {
     bool mapping = submap_id.trajectory_id == 0 && node_id.trajectory_id == 0;
-    bool initializing = submap_id.submap_index <= 1 && node_id.trajectory_id == 0;
-    bool log = mapping || initializing || submap_id.trajectory_id != node_id.trajectory_id;
+    bool inter_submap_constraint = submap_id.trajectory_id != node_id.trajectory_id;
```
Comment (johuber) [inline cartographer/mapping/internal/constraints/constraint_builder_3d.cc]: const

File: cartographer/mapping/internal/constraints/constraint_builder_3d.cc [hunk 7]
```diff
@@ -228,8 +228,8 @@ void ConstraintBuilder3D::ComputeConstraint(
     }
   } else {
     bool mapping = submap_id.trajectory_id == 0 && node_id.trajectory_id == 0;
```
Comment (johuber) [inline cartographer/mapping/internal/constraints/constraint_builder_3d.cc]: const

File: cartographer/mapping/internal/constraints/constraint_builder_3d.cc [hunk 8]
```diff
@@ -228,8 +228,8 @@ void ConstraintBuilder3D::ComputeConstraint(
...
-    bool initializing = submap_id.submap_index <= 1 && node_id.trajectory_id == 0;
-    bool log = mapping || initializing || submap_id.trajectory_id != node_id.trajectory_id;
+    bool inter_submap_constraint = submap_id.trajectory_id != node_id.trajectory_id;
+    bool log = mapping || inter_submap_constraint;
```
Comment (johuber) [inline cartographer/mapping/internal/constraints/constraint_builder_3d.cc]: const

**Review by johuber — CHANGES_REQUESTED** [review]

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 9]
```diff
@@ -289,6 +290,15 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
                                            constant_data, global_node_pose,
                                            global_submap_pose);
   } else if (maybe_add_global_constraint) {
+    const float distance = (global_node_pose.translation()- global_submap_pose.translation()).norm();
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: missing space before `-`

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 10]
```diff
@@ -289,6 +290,15 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
...
+    if (distance > options_.constraint_builder_options().max_constraint_distance()) {
+        LOG(INFO) << "[SKIPPED] Skipped searching from " << node_id.trajectory_id
+            << ", node " << node_id.node_index << " to subamp " << submap_id.submap_index
+            << " because of distance: " << distance;
+      return;
+    }
+    LOG(INFO) << "Searching global constraint from trajectory " << node_id.trajectory_id
+        << ", node " << node_id.node_index << " to subamp " << submap_id.submap_index;
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: typo in `subamp`

File: cartographer/mapping/internal/3d/pose_graph_3d.cc [hunk 11]
```diff
@@ -289,6 +290,15 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
                                            constant_data, global_node_pose,
                                            global_submap_pose);
   } else if (maybe_add_global_constraint) {
+    const float distance = (global_node_pose.translation()- global_submap_pose.translation()).norm();
+    if (distance > options_.constraint_builder_options().max_constraint_distance()) {
+        LOG(INFO) << "[SKIPPED] Skipped searching from " << node_id.trajectory_id
+            << ", node " << node_id.node_index << " to subamp " << submap_id.submap_index
```
Comment (johuber) [inline cartographer/mapping/internal/3d/pose_graph_3d.cc]: typo in `subamp`

File: cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.h [hunk 12]
```diff
@@ -115,7 +116,7 @@ class FastCorrelativeScanMatcher3D {
       const sensor::PointCloud& point_cloud,
       const Eigen::VectorXf& rotational_scan_matcher_histogram,
       const Eigen::Quaterniond& gravity_alignment, float min_score,
-      const bool log = false) const;
+      const bool full = false, const bool log = false) const;
```
Comment (johuber) [inline cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.h:119]: does the default argument make sense here? It seems like you set it anywhere and it has to be added everywhere anyway

File: cartographer/mapping/internal/constraints/constraint_builder_3d.cc [hunk 13]
```diff
@@ -224,18 +229,25 @@ void ConstraintBuilder3D::ComputeConstraint(
...
+      LOG(INFO) << "[DROPPED] " << search_type << ": Dropped from "
+          << node_id.trajectory_id << ", node " << node_id.node_index
+          << " to " << submap_id.submap_index;
       return;
     }
   } else {
-    bool mapping = submap_id.trajectory_id == 0 && node_id.trajectory_id == 0;
-    bool initializing = submap_id.submap_index <= 1 && node_id.trajectory_id == 0;
-    bool log = mapping || initializing || submap_id.trajectory_id != node_id.trajectory_id;
+    const bool mapping = submap_id.trajectory_id == 0 && node_id.trajectory_id == 0;
```
Comment (johuber) [inline cartographer/mapping/internal/constraints/constraint_builder_3d.cc:238]: why is this called mapping? looks like something else to me..

**Review by johuber — CHANGES_REQUESTED** [review]
Mostly cosmetics

File: cartographer/mapping/internal/constraints/constraint_builder_3d.cc (hunk 13 above)
Comment (AlexReimann) [reply cartographer/mapping/internal/constraints/constraint_builder_3d.cc:238]: Searching constraints from `trajectory_id == 0` to `trajectory_id == 0`.
-> No other trajectory before that, so it is mapping 

What does it look like to you?

Comment (johuber) [reply cartographer/mapping/internal/constraints/constraint_builder_3d.cc:238]: can't it also be mapping if `trajectory_id  == 100`? wouldn't it be more related to `submap_id.trajectory_id == node_id.trajectory_id`? or maybe I fully misunderstand..

Comment (AlexReimann) [reply cartographer/mapping/internal/constraints/constraint_builder_3d.cc:238]: `submap_id.trajectory_id == node_id.trajectory_id` is inter trajectory constraints, that's happening all the time.
Yeah in theory someone could map with a different starting trajectory, but for us it's not the case (for now)

File: cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.h (hunk 12 above)
Comment (AlexReimann) [reply cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.h:119]: Original upstream doesn't have the parameters, is backwards compatibility with upstream to not get conflicts / have less changes.

**Review by johuber — APPROVED** [review]

**Pull request merged**